File: CTCI/Interview Questions/C++/2.3_delete_middle_node.cpp
// Implement an algorithm to delete a node in the midddle (i.e. any node but the first and last node, not necessarily the exact middle) of a singly linked list, given only access to that node

// Page 63

#include <iostream>
#include <vector>
#include <ctime>
using namespace std;

struct Node{
    int data;
    Node * next = NULL;

    void set_next(Node * next_node){
        next = next_node;
    }
    Node * get_next(){
        return next;
    }
    void set_data(int new_data){
        data = new_data;
    }
    int get_data(){
        return data;
    }
};

class LinkedList{
private:
    Node * head = NULL;
    int size = 0;
public:
    void append(int data){
        Node * new_node = new Node;
        new_node -> data = data;
        if(head == NULL)
            head = new_node;
        else{
            Node * current = head;
            while(current -> next)
                current = current -> next;
            current -> next = new_node;
        }
        size++;
    }
    void fill(vector <int> elements){
        for(auto element : elements)
            append(element);
    }
    void print(){
        Node * current = head;
        while(current){
            cout << current -> data << " ";
            current = current -> next;
        }
        cout << endl;
    }
    int get_size(){
        return size;
    }
    Node * get_head(){
        return head;
    }
};

vector <int> create_random_vect(int num_elements){
    vector <int> return_vect;
    for(int i = 1; i < num_elements + 1; i++)
        return_vect.push_back(rand() % num_elements + 1);
    return return_vect;
}

// Solution 1
// Set mid's data to next data, and set mid's next to next's next
// Time Complexity: O(n), Space Complexity: O(1)
void delete_middle_node(Node * mid_node){
    if(mid_node == NULL || mid_node -> next == NULL) return;
    mid_node -> data = mid_node -> next -> data;
    mid_node -> next = mid_node -> next -> next;
}


int main(void){
    srand(time(NULL)); // Using time as seed for RNG
    int num_elements = 10;
    vector <int> random_vect = create_random_vect(num_elements);
    LinkedList * ll = new LinkedList;
    (*ll).fill(random_vect);
    (*ll).print();

    Node * mid_node = (*ll).get_head();
    int i = 1;
    while(i < num_elements / 2){
        i++;
        mid_node = mid_node -> next;
    }
    cout << "Removing middle node (" << mid_node -> data << ")" << endl;
    delete_middle_node(mid_node);
    (*ll).print();
}

File: CTCI/Interview Questions/Python/2.3_delete_middle_node.py
# Implement an algorithm to delete a node in the midddle (i.e. any node but the first and last node, not necessarily the exact middle) of a singly linked list, given only access to that node

# Page 63

import random

class Node:
    def __init__(self, data):
        self.data = data
        self.next = None

class LinkedList:
    def __init__(self):
        self.head = None
        self.size = 0

    def fill(self, elements):
        for element in elements:
            self.append(element)

    def append(self, data):
        if self.head is None:
            self.head = Node(data)
        else:
            current = self.head
            while current.next:
                current = current.next
            current.next = Node(data)
        self.size += 1

    def __str__(self):
        output_list = []
        current = self.head
        while current:
            output_list.append(current.data)
            current = current.next
        return ' '.join([str(ele) for ele in output_list])

# Solution 1
# Starting from the middle (because that's all we have access to), copy the data from the next node (if it exists) and set it to the node before it.
# Time Complexity: O(n), Space Complexity: O(1)
def delete_middle_node(mid_node):
    while mid_node.next:
        mid_node.data = mid_node.next.data
        if mid_node.next.next is None:
            mid_node.next = None
            break
        mid_node = mid_node.next

num_nodes = 10
ll = LinkedList()
ll.fill([random.randint(1, num_nodes) for i in range(num_nodes)])
print(ll)

mid_node = ll.head
i = 1
while i < num_nodes//2:
    i += 1
    mid_node = mid_node.next
print('Removing middle node ({})'.format(mid_node.data))
delete_middle_node(mid_node)
print(ll)
